File: Python/MP11_ImageSmoothing.py
import cv2 as cv
import numpy as np    
from PIL import Image  

# img = cv.imread("Lenna.jpg", cv.IMREAD_COLOR)
# imgPIL = Image.open('Lenna.jpg')

img = cv.imread("bird_small.jpg", cv.IMREAD_COLOR)
imgPIL = Image.open('bird_small.jpg')

class imageSize:
    height, width, _ = img.shape

def ImageSmoothing(imgPIL, mask):
    if mask == 3:   
        offset = 1
    if mask == 5:   
        offset = 2
    if mask == 7:   
        offset = 3
    if mask == 9:   
        offset = 4        
    SmoothedImage = Image.new(imgPIL.mode, imgPIL.size)
    for x in range(offset, imageSize.width - offset):
        for y in range(offset, imageSize.height - offset):
            #Lay gia tri cac diem anh tai vi tri (x,y)
            Rs = 0
            Gs = 0
            Bs = 0
            for i in range(x - offset,x + offset + 1):
                for j in range(y - offset,y + offset + 1):
                    #Lay gia tri cac diem anh tai vi tri (x,y)
                    R, G, B = imgPIL.getpixel((i,j))
                    Rs += R
                    Gs += G
                    Bs += B
            K = mask * mask
            Rs = np.uint8(Rs/K)   
            Bs = np.uint8(Bs/K)
            Gs = np.uint8(Gs/K)

            SmoothedImage.putpixel((x,y),(Bs, Gs, Rs))

    SmoothedImage3x3_CV = np.array(SmoothedImage)
    return SmoothedImage3x3_CV


img_3x3 = ImageSmoothing(imgPIL, 3)
img_5x5 = ImageSmoothing(imgPIL, 5)
img_7x7 = ImageSmoothing(imgPIL, 7)
img_9x9 = ImageSmoothing(imgPIL, 9)

cv.imshow('Anh Goc', img)
cv.imshow('Anh lam muot 3x3', img_3x3)
cv.imshow('Anh lam muot 5x5', img_5x5)
cv.imshow('Anh lam muot 7x7', img_7x7)
cv.imshow('Anh lam muot 9x9', img_9x9)
cv.waitKey(0)
cv.destroyAllWindows()

File: C#.NET/MP11_ImageSmoothing.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace SmoothImage
{
    public partial class ImageSmoothing : Form
    {
        public ImageSmoothing()
        {
            InitializeComponent();
        }

        private void label2_Click(object sender, EventArgs e)
        {

        }

        private void Form1_Load(object sender, EventArgs e)
        {
            Bitmap hinhGoc = new Bitmap(@"E:\Nam_3\ThiGiacMay\Images\Lenna.jpg");
            // Load anh vao picbox
            picBoxHinhGoc.Image = hinhGoc;
            picBox_3x3.Image = ColorImageSmoothing(hinhGoc, 3);
            picBox_5x5.Image = ColorImageSmoothing(hinhGoc, 5);
            picBox_7x7.Image = ColorImageSmoothing(hinhGoc, 7);
            picBox_9x9.Image = ColorImageSmoothing(hinhGoc, 9);

        }
        public Bitmap ColorImageSmoothing(Bitmap hinhGoc, int mask)
        {
            Bitmap SmoothedImage = new Bitmap(hinhGoc.Width, hinhGoc.Height);
            // Quet cac diem anh trong hinh goc RGB
            // Do mat na 3x3 nen bo qua vien ngoai cua anh
            int offset = 0;
            if (mask == 3) offset = 1; 
            if (mask == 5) offset = 2; 
            if (mask == 7) offset = 3; 
            if (mask == 9) offset = 4;

            for (int x = offset; x < hinhGoc.Width - offset; x++)
                for(int y = offset; y < hinhGoc.Height - offset; y++)
                {
                    //Cac bien de chua gia tri cong don cac diem anh
                    int Rs = 0, Gs = 0, Bs = 0;

                    //Quet cac diem anh trong mat na
                    for (int i = x - offset; i <= x + offset; i++)
                        for (int j = y - offset; j <= y + offset; j++)
                        {
                            //Lay thong tin mau RGB tai diem anh trong mat na
                            //tai vi tri (i;j)
                            Color color = hinhGoc.GetPixel(i, j);
                            byte R = color.R;
                            byte G = color.G;
                            byte B = color.B;
                            //Cong don tat ca cac diem ah do cho moi kenh
                            Rs += R;
                            Gs += G;
                            Bs += B;
                        }
                    //Ket thuc quet va cong don diem anh trong mat na
                    //Bat dau tinh trung binh cong cho moi kenh theo cong thuc 6.6-1 trong sach
                    //cho tung kenh mau R,G,B
                    int K = mask * mask;
                    Rs = (int)(Rs / K);
                    Gs = (int)(Gs / K);
                    Bs = (int)(Bs / K);

                    SmoothedImage.SetPixel(x, y, Color.FromArgb(Rs, Gs, Bs));
                }
            return SmoothedImage;
        }

    }
}
